### PR TITLE
Car sensors and movement

### LINKED ISSUE #2: Car sensors
Car should have sensors for walls' positions.

### LINKED ISSUE #3: Car movement
Car should move in 2D space based on inputs like acceleration and turn force.

### PR DESCRIPTION
**Pull request by jakubdabek**
Closes #2

### CODE REVIEW

File: src/model/geom/track.py [hunk 1]
```diff
@@ -41,12 +42,50 @@ def segment_walls(self, segment_id: SegmentId) -> List[Wall]:
...
-        raise RuntimeError("couldn't find any walls in any of the segments")
+                for i, sensor in sensors_to_check[:]:
+                    if (d := sensor.check_distance(wall)) is not None:
+                        distances[i] = d
+                        del sensors_to_check[i]
+
+        if any(d is None for d in distances):
+            raise RuntimeError("couldn't find any walls in any of the segments")
+
+        return cast(List[float], distances)
```
Comment (ErykKrupa) [inline src/model/geom/track.py]: In my humble opinion, it isn't the true optimalization you said about in todo. We still iterate through all segments and all sensors, but now in one method, not in one method per sensor. What's worse, we make copy of sensors_to_check for every wall. I believe, code below should resolve both issues.
```
sensors_to_check: List[Tuple[int, Sensor]] = list(enumerate(sensors))
sensors_to_remove: List[int] = []
distances: List[Optional[float]] = [None] * len(sensors_to_check)
distances_to_count = len(distances)
for segment_id in self._spread_segment_ids(active_segment):
    for wall in self.segment_walls(segment_id):
        for i, sensor in sensors_to_check:
            if (d := sensor.check_distance(wall)) is not None:
                distances[i] = d
                sensors_to_remove.append(i)
                distances_to_count -= 1
                if distances_to_count == 0:
                    return cast(List[float], distances)
        for _ in range(len(sensors_to_remove)):
            del sensors_to_check[sensors_to_remove.pop()]
raise RuntimeError("couldn't find any walls in any of the segments")
```

File: src/model/geom/track.py [hunk 2]
```diff
@@ -41,12 +42,50 @@ def segment_walls(self, segment_id: SegmentId) -> List[Wall]:
...
-        for segment_id in utils.spread_int(active_segment, 0, len(self.segments)):
+    def _spread_segment_ids(self, segment_id: SegmentId) -> Iterable[SegmentId]:
+        return utils.spread_int(segment_id, 0, len(self.segments))
+
+    def sense_closest(
+        self, sensors: List[Sensor], active_segment: SegmentId
+    ) -> List[float]:
+        """Returns the distance to the closest wall of the track sensed with the given sensor."""
+        # TODO?: edge cases when a later segment has
+        #  the closest wall (not sure if that ever happens)
```
Comment (ErykKrupa) [inline src/model/geom/track.py:53]: I was thinking about it before and I can't imagine this situation, but also I'm not able to present mathematical proof.

File: src/model/geom/track.py [hunk 3]
```diff
@@ -41,12 +42,50 @@ def segment_walls(self, segment_id: SegmentId) -> List[Wall]:
...
+            if not any(current_bbx.contains_point(p) for p in shape):
+                found_limit[i % 2] = True
+                if all(found_limit):
+                    return False
+        return False
+
+    def update_active(self, active_segment: SegmentId, center: Point) -> SegmentId:
+        """Returns segment currently containing the given point
+        assuming it was moved by a small amount and is still inside the track.
+        """
```
Comment (KamilKrol5) [inline src/model/geom/track.py]:
```suggestion
        """
        Returns segment currently containing the given point 
        assuming it was moved by a small amount and is still inside the track.
        """
```

Comment (jakubdabek) [reply src/model/geom/track.py]: Done :)

File: src/model/geom/track.py (hunk 1 above)
Comment (jakubdabek) [reply src/model/geom/track.py]: Current version should have this resolved. Please check again.

**Review by ErykKrupa — APPROVED** [review]

**Review by KamilKrol5 — APPROVED** [review]

**Review by BartoszRodak — APPROVED** [review]

**Pull request merged**